### ISSUE TITLE
Overlay colors according to categorical variables

### ISSUE THREAD

**Issue by kikegoni** [iSEE/iSEE#405]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by kikegoni** [comment]
Hey all,

First of all, thank you for sharing iSEE, I am finding it super useful to visualize and share data. 

I am visualizing my cells in a UMAP that belong to a particular BATCH_X against the remaining. I have ~2000 cells that belong to BATCH_X and ~50000 that do not. The problem is that the cells that don't belong to BATCH_X are obscuring the ones that belong, so I am only seeing a one color plot: 

``` r
labels <- sce@colData$batch_X
UMAP <- sce@reducedDims$UMAP
plot(UMAP,  pch = 20, cex=0.4 , col = ifelse(labels==T, "green","grey"))
```

![image](https://user-images.githubusercontent.com/35218702/80059925-d202af00-855f-11ea-9a08-388bb9ff9b7e.png)

For this reason, I would like to overlay those cells that belong to BATCH_X like: 

![image](https://user-images.githubusercontent.com/35218702/80060033-26a62a00-8560-11ea-86c1-d8bab234f6e7.png)

Which can be done simply with: 

```r
plot(UMAP[labels==F,], pch = 20, cex=0.4, col = "grey")
points(UMAP[labels==T,], pch = 20, cex=0.4, col = "green")
```

Do you have any insight about this? The idea would be to tell the colormap that cells belonging to one condition should overlay cells belonging to other conditions. 

Thanks for any help;)

**Comment by kevinrue** [comment]
Hi @kikegoni 

Thanks for your feedback and your kind words about `iSEE`

The code that you posted isn't coming from `iSEE`, nor are the plots; what you describe is not an issue with `iSEE` and doesn't belong here; instead I recommend the [Bioconductor support site](https://support.bioconductor.org/).

That said, `iSEE` does include a neat instructional feature that can help you solve your problem.
Namely, the `iSEE` application gives users the code that it uses to deal address this issue.

- Run the example code is `?iSEE`.
- In the "Reduced dimension plot" panel, in the "Visual parameters" box, in the "Points" menu, turn on the downsampling.
- In the "Diagnostics" menu, click "Extract the R code"
- In the "Reduced dimension plot" section, look at the code.

You will see the chunk of code:

```
# Randomize data points to avoid a data set bias during the downsampling
set.seed(100);
plot.data <- plot.data[sample(nrow(plot.data)), , drop=FALSE];
```

The bottom line here is that shuffling the data points before plotting will increase the chances that some of them are visible in the plot.
Essentially, it removes the bias whereby the cells of a given batch are all plotted first, and subsequently buried under cells from all subsequent batches.

In your example, that would look like:
```
UMAP <- UMAP[sample(nrow(UMAP)), , drop=FALSE];
```

Alternatively (I only recommend this for very small batches), you could move all points of the batch of interest at the end of the dataset, so that they are all plotted on top of all other batches.

In your example, that would look like
```
UMAP <- rbind(UMAP[labels==FALSE,,drop=FALSE], UMAP[labels==TRUE,,drop=FALSE])
```

**Comment by kevinrue** [comment]
Note to self: This made me notice that we don't shuffling data points when downsampling is OFF, which actually produces this kind of issue.

**Comment by kikegoni** [comment]
Hey @kevinrue,

Thanks a lot for your answer. This was exactly what I was looking for. As you said actually the data points shuffling is only happening when downsampling is ON, so are you planning to edit it also for the cases when downsampling is OFF?

**Comment by LTLA** [comment]
> are you planning to edit it also for the cases when downsampling is OFF?

This is already the case. Shuffling is performed all the time in the latest version.

> But in the case that we have a boolean colData argument (my batch_X) and we want the data points that are TRUE for that colData to overlay over the ones that are FALSE, I have not been able to solve this with your platform.

We have no support for this in the default panels. It is implemented for special panel types (e.g., the MA and volcano plots in **iSEEu**, where DE genes are shown on top), but for all other panels, the color and order of points are independent considerations.

While it would not be too difficult to add, faceting by your factor of interest is a more general approach that works for any number of levels and avoids excessive overlaps.

**Comment by kikegoni** [comment]
Hey @LTLA @kevinrue ,

Thanks a lot for your helpful answers. That will be enough and more than appreciated. 

Anyway I have realized that 
```r
BiocManager::install("iSEE")
```

Is not installing the latest (2.0) version, Instead still install 1.6.1 version

**Comment by kikegoni** [comment]
Actually I have just installed your latest version with github and I have seen that the data shuffle is not happenning when downsampling is OFF. I am using your iSEE version 2.1.2 and what I see when I set downsampling to OFF is:

```r
red.dim <- reducedDim(se, 1);
plot.data <- data.frame(X=red.dim[, 1], Y=red.dim[, 2], row.names=colnames(se));
plot.data$ColorBy <- colData(se)[, "chem"];
plot.data <- subset(plot.data, !is.na(X) & !is.na(Y));

# Saving data for transmission
all_coordinates[['redDimPlot1']] <- plot.data

# Creating the plot
ggplot() +
    geom_point(aes(x=X, y=Y, color=ColorBy), alpha=1, plot.data, size=1) +
    labs(x="Dimension 1", y="Dimension 2", color="chem", title="(1) UMAP") +
    coord_cartesian(xlim=range(plot.data$X, na.rm=TRUE),
        ylim=range(plot.data$Y, na.rm=TRUE), expand=TRUE) +
    scale_color_manual(values=colDataColorMap(colormap, "chem", discrete=TRUE)(2), na.value='grey50', drop=FALSE) +
    scale_fill_manual(values=colDataColorMap(colormap, "chem", discrete=TRUE)(2), na.value='grey50', drop=FALSE) +
    theme_bw() +
    theme(legend.position='bottom', legend.box='vertical', legend.text=element_text(size=9), legend.title=element_text(size=11),
            axis.text=element_text(size=10), axis.title=element_text(size=12), title=element_text(size=12))
```

And when downsampling is ON:
```r
red.dim <- reducedDim(se, 1);
plot.data <- data.frame(X=red.dim[, 1], Y=red.dim[, 2], row.names=colnames(se));
plot.data$ColorBy <- colData(se)[, "chem"];
plot.data <- subset(plot.data, !is.na(X) & !is.na(Y));

# Saving data for transmission
all_coordinates[['redDimPlot1']] <- plot.data

# Creating the plot
plot.data.pre <- plot.data;
# Randomize data points to avoid a data set bias during the downsampling
set.seed(100);
plot.data <- plot.data[sample(nrow(plot.data)), , drop=FALSE];
plot.data <- subset(plot.data, subsetPointsByGrid(X, Y, resolution=200));

ggplot() +
    geom_point(aes(x=X, y=Y, color=ColorBy), alpha=1, plot.data, size=1) +
    labs(x="Dimension 1", y="Dimension 2", color="chem", title="(1) UMAP") +
    coord_cartesian(xlim=range(plot.data.pre$X, na.rm=TRUE),
        ylim=range(plot.data.pre$Y, na.rm=TRUE), expand=TRUE) +
    scale_color_manual(values=colDataColorMap(colormap, "chem", discrete=TRUE)(2), na.value='grey50', drop=FALSE) +
    scale_fill_manual(values=colDataColorMap(colormap, "chem", discrete=TRUE)(2), na.value='grey50', drop=FALSE) +
    theme_bw() +
    theme(legend.position='bottom', legend.box='vertical', legend.text=element_text(size=9), legend.title=element_text(size=11),
            axis.text=element_text(size=10), axis.title=element_text(size=12), title=element_text(size=12))
```

**Comment by LTLA** [comment]
Check your version. For the current release (2.0.0), I get:

```r
################################################################################
## Reduced dimension plot 1
################################################################################


red.dim <- reducedDim(se, "PCA");
plot.data <- data.frame(X=red.dim[, 1], Y=red.dim[, 2], row.names=colnames(se));

# Avoid visual biases from default ordering by shuffling the points
set.seed(379);
plot.data <- plot.data[sample(nrow(plot.data)),,drop=FALSE];

dot.plot <- ggplot() +
    geom_point(aes(x=X, y=Y), alpha=1, plot.data, color='#000000', size=1) +
    labs(x="Dimension 1", y="Dimension 2", title="PCA") +
    coord_cartesian(xlim=range(plot.data$X, na.rm=TRUE),
        ylim=range(plot.data$Y, na.rm=TRUE), expand=TRUE) +
    theme_bw() +
    theme(legend.position='bottom', legend.box='vertical', legend.text=element_text(size=9), legend.title=element_text(size=11),
            axis.text=element_text(size=10), axis.title=element_text(size=12), title=element_text(size=12))
```

**Comment by kikegoni** [comment]
Hey @LTLA ,

How do you install version 2.0.0?

**Comment by LTLA** [comment]
> How do you install version 2.0.0?

Follow the Bioconductor instructions carefully: in particular, the `(version "4.0")` is **not** optional.

> And from github I get version 2.1.2 with

2.1.2 has the same behavior as 2.0.0. I daresay you are not restarting R after package installation.

**Comment by kikegoni** [comment]
Ahhhh!!! yes, sorry, both of your assumptions were correct!! Now is working! Thank you very much for your help
